### PR TITLE
Support multidimensional (nested) arrays

### PR DESCRIPTION
**Pull request by saifbechan**
In our application during registration a user can select multiple option. These options are stored in a multidimensional array. We want to be able to get the value of each field in the given path.

Take for example the following document:

```
{
  field_1: "tagValue1",
  field_2: [
    { label : "label_1", value: "value_1" },
    { label : "label_2", value: "value_2" },
    { label : "label_3", value: "value_3" },
  ]
}
```

From this we want to extract `field_1` and every `value` of `field_2`. The end result will look like this:

```
{
  tags: [
    { name: "tagValue1", status: "active" },
    { name: "value_1", status: "active" },
    { name: "value_2", status: "active" },
    { name: "value_3", status: "active" },
  ],
}
```

The value for `memberTags` will look like this: `["tag_data.field_1", { path: "tag_data.field_2", key: "value" }]`.

### CODE REVIEW

**Comment by randykinne** [comment]
Hey! Thank you for the pr, I've forwarded it to the relevant teams. Unfortunately I don't have access to request specific reviewers in this repo so it's up to them

**Comment by barticus** [comment]
Hi @saifbechan ,

I am on parental leave until the new year, so I'll pass this along to my colleague but I have some thoughts.

Firstly, thank you for contributing! We've been looking at expanding the config formats for different types (such as [outlined here](https://github.com/mailchimp/Firebase/issues/43)), to support greater feature sets, and this looks in line with that.

For the overall approach, I think you're on the right track. However, I'd prefer:

1. The config property names to be explicit, e.g `path` to be `mailchimpTagPath` and `key` to be `keySelector`
2. The implementation should not be tied so tightly to multidimensional arrays, e.g. 
    ```
    ["tag_data.field_1", { mailchimpTagPath: "tag_data.field_2" }]
    ```
    should be equivalent to 
    ```
    ["tag_data.field_1", "tag_data.field_2" ]
    ```

I'll add some inline comments that should help with this too.

File: functions/index.js [hunk 1]
```diff
@@ -159,9 +159,16 @@ exports.memberTagsHandler = functions.handler.firestore.document
...
       const newDoc = event && event.after && event.after.data();
 
+      // Check if we are dealing with a multidimensional array
+      const isMultidimensional = (tag) =>
+        typeof tag === "object" && typeof tag.path !== "undefined" && typeof tag.key !== "undefined"
+
       // Retrieves subscriber tags before/after write event
       const getTagsFromEventSnapshot = snapshot => tagsConfig.memberTags.reduce((acc, tag) => {
-        const tags = _.get(snapshot, tag);
+        const tags = isMultidimensional(tag)
```
Comment (barticus) [inline functions/index.js:168]: Please update to

```
const tags = isObjectConfig(tag)
          ? _.get(snapshot, tag.tagPath).map((parent) => typeof parent === "object" && typeof tag.keySelector !== "undefined" ? _.get(parent, tag.keySelector) : parent)
          : _.get(snapshot, tag);
```

File: functions/tests/memberTagsHandler.test.js [hunk 2]
```diff
@@ -183,6 +183,50 @@ describe("memberTagsHandler", () => {
     );
   });
 
+  it("should set tags from multidimensional nested config for new user", async () => {
```
Comment (barticus) [inline functions/tests/memberTagsHandler.test.js:186]: Once the other comments are resolved, please add an additional test for when keySelector is not provided but object format is.

**Review by barticus — CHANGES_REQUESTED** [review]

**Comment by barticus** [comment]
Can you please also update the destination branch to [release/0.5.0](https://github.com/mailchimp/Firebase/tree/release/0.5.0)?

File: functions/index.js (hunk 1 above)
Comment (barticus) [reply functions/index.js:168]: Please extend on this to check that the return value when using tagPath is an array before calling map

**Comment by saifbechan** [comment]
@barticus thanks for the feedback. Will work on these changes soon.

**Comment by barticus** [comment]
Hi @saifbechan , I have implemented your requested changes as part of #53 . Happy for you to provide feedback, please let me know if this solution works for you, but it should as I modelled it from your problem description.

Regards.

**Comment by barticus** [comment]
Hi @saifbechan , after some consideration I altered the implementation for this sort of problem. Instead of requiring additional configuration elements, I've integration a [JMESPath](https://jmespath.org/) library which means, for your earlier example document:

```
{
  field_1: "tagValue1",
  field_2: [
    { label : "label_1", value: "value_1" },
    { label : "label_2", value: "value_2" },
    { label : "label_3", value: "value_3" },
  ]
}
```

In order to extract the tag values of `["value_1", "value_2", "value_3"]`, your memberTagsConfig will only need to be:

```
{ "memberTags": ["field_2[*].value"], "subscriberEmail": "emailAddress" }
```

This is implemented in the newly launched 0.5.0 release from PR #58

**Pull request closed without merging**